transport.go

```diff
@@ -14,221 +13,196 @@ type UnwrapErrorFunc func(nxt DecodeNext) (error, error)
+	log            LogInterface
+	wrapError      WrapErrorFunc
+	startOnce      sync.Once
+	writeCh        chan []byte
+	readByteCh     chan struct{}
+	decodeCh       chan interface{}
+	readerResultCh chan interface{}
+	writerResultCh chan error
+	stopCh         chan struct{}
 }
 
-func NewTransport(c net.Conn, l LogFactory, wef WrapErrorFunc) *Transport {
-	mh := codec.MsgpackHandle{WriteExt: true}
-
-	buf := new(bytes.Buffer)
-	ret := &Transport{
-		mh:        &mh,
-		cpkg:      NewconPackage(c, &mh),
-		buf:       buf,
-		enc:       codec.NewEncoder(buf, &mh),
-		mutex:     new(sync.Mutex),
-		rdlck:     new(sync.Mutex),
-		wrlck:     new(sync.Mutex),
-		wrapError: wef,
-	}
+func NewTransport(c net.Conn, l LogFactory, wef WrapErrorFunc) Transporter {
+	cdec := newConnDecoder(c)
 	if l == nil {
 		l = NewSimpleLogFactory(nil, nil)
 	}
-	log := l.NewLog(ret.cpkg.GetRemoteAddr())
-	ret.log = log
-	ret.dispatcher = NewDispatch(ret, log, wef)
-	ret.packetizer = NewPacketizer(ret.dispatcher, ret)
+	log := l.NewLog(cdec.RemoteAddr())
+	byteEncoder := newFramedMsgpackEncoder()
+
+	ret := &transport{
+		Locker:         new(sync.Mutex),
+		ByteEncoder:    byteEncoder,
+		cdec:           cdec,
+		log:            log,
+		wrapError:      wef,
+		writeCh:        make(chan []byte),
+		readByteCh:     make(chan struct{}),
+		decodeCh:       make(chan interface{}),
+		readerResultCh: make(chan interface{}),
+		writerResultCh: make(chan error),
+		stopCh:         make(chan struct{}),
+	}
+	ret.dispatcher = newDispatch(ret.writeCh, ret.writerResultCh, log, wef)
 	return ret
 }
 
-func (t *Transport) ReadLock()   { t.rdlck.Lock() }
-func (t *Transport) ReadUnlock() { t.rdlck.Unlock() }
-
-func (t *Transport) encodeToBytes(i interface{}) (v []byte, err error) {
-	if err = t.enc.Encode(i); err != nil {
-		return
+func (t *transport) IsConnected() bool {
+	select {
+	case <-t.stopCh:
+		return false
+	default:
+		return true
 	}
-	v, _ = ioutil.ReadAll(t.buf)
-	return
 }
 
-func (t *Transport) run2() (err error) {
-	err = t.packetizer.Packetize()
-	t.handlePacketizerFailure(err)
+func (t *transport) Run() (err error) {
+	if !t.IsConnected() {
+		return DisconnectedError{}
+	}
+	t.startOnce.Do(func() {
+		err = t.run2()
+	})
 	return
 }
 
-func (t *Transport) handlePacketizerFailure(err error) {
-	// For now, just throw everything away.  Eventually we might
-	// want to make a plan for reconnecting.
-	t.mutex.Lock()
-	t.running = false
-	t.dispatcher.Reset()
-	t.dispatcher = nil
-	t.packetizer.Clear()
-	t.packetizer = nil
-	t.cpkg.Close()
-	t.cpkg = nil
-	t.mutex.Unlock()
-	// NOTE: The logging implementation can be anything. In particular, it
-	// might try to send logs over this transport, which would take the mutex
-	// again. We *must not* call this while we hold the lock. (Yes, we figured
-	// this out by deadlocking ourselves :p)
+func (t *transport) run2() (err error) {
+	// Initialize transport loops
+	readerDone := runInBg(t.readerLoop)
```

Comment:
Because I only found out about `sync.WaitGroup` about an hour ago, and by then this seemed prettier with less boilerplate.